### PR TITLE
[FIRRTL] Block constant propagation through DontTouchAnnotation

### PR DESCRIPTION
**Pull request by prithayan**
The `DontTouch` annotation should block constant propagation through it, according to the Scala [FIRRTL implementation](https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/ConstantPropagation.scala) 
Mimic this behavior in `IMConstProp` pass.
1. Check if the value has a `DontTouch` annotaiton, whenever updating its lattice value.
2. Mark it as `overdefined`, such that it blocks any constant propagation through it.

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h [hunk 1]
```diff
@@ -68,6 +68,9 @@ class AnnotationSet {
   forPort(Operation *module, size_t portNo,
           SmallVectorImpl<NamedAttribute> &otherAttributes);
 
+  /// Get an annotation set for the specified value.
+  static AnnotationSet getAnnotationSet(Value v);
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:72]: The naming seems verbose to me. However, I don't know what is a good naming convention. `get` seems to align better with my understanding of MLIR conventions.

```suggestion
  /// Get an annotation set for the specified value.
  static AnnotationSet get(Value v);
```

**Review by seldridge — APPROVED** [review]
Looks solid to me. 

Can you update the documentation for `DontTouchAnnotation` [here](https://github.com/llvm/circt/blob/main/docs/FIRRTLAnnotations.md#donttouchannotation) to indicate that this is an optimization barrier and not the weaker definition of "something that will exist in the output"?

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h (hunk 1 above)
Comment (prithayan) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h]: Thanks, updated the name.

**Comment by prithayan** [comment]
Thanks @seldridge , Updated the doc.

**Review by seldridge — APPROVED** [review]

**Comment by seldridge** [comment]
In a follow-up, it would be reasonable to make sure that we have the same behavior for all the dont touch tests here: https://github.com/chipsalliance/firrtl/blob/master/src/test/scala/firrtlTests/ConstantPropagationTests.scala. This is basically just enumerating across ports, registers, nodes, wires, etc. and making sure that it does the right thing in all situations.

**Comment by prithayan** [comment]
Thanks @seldridge , added rest of the test cases also. Merging this now.

**Pull request merged**